### PR TITLE
Add high order grad support for Some operator

### PR DESCRIPTION
**Pull request by caogang**
For high order grad support in new-style function, solving the issue #1483 . I have finished the features:
- [x] Threshold
- [x] ReLU
- [x] Sqrt
- [x] Norm

### CODE REVIEW

File: torch/autograd/_functions/reduce.py
Comment (fmassa) [inline torch/autograd/_functions/reduce.py]: This is not right, you are unwrapping the Variable, which means it is losing its history.

File: torch/autograd/_functions/reduce.py
Comment (fmassa) [inline torch/autograd/_functions/reduce.py]: Same thing here.
you should instead keep the Variable, but expand it to the same size as grad input before multiplication

File: torch/autograd/_functions/reduce.py
Comment (apaszke) [inline torch/autograd/_functions/reduce.py]: This indexing is unnecessary now

File: torch/autograd/_functions/reduce.py
Comment (apaszke) [inline torch/autograd/_functions/reduce.py]: Same here

File: torch/autograd/_functions/reduce.py
Comment (apaszke) [inline torch/autograd/_functions/reduce.py]: Can you please change `norm_type` to `p`, so that it matches the name used in `Variable`?

File: torch/nn/_functions/thnn/activation.py
Comment (apaszke) [inline torch/nn/_functions/thnn/activation.py]: Can you add a TODO that says it should be removed once we add support for Variable backward in auto-generated THNN functions? While I'll accept this change, because I want this feature to be available quickly, I don't want this to be our long term solution.

File: torch/nn/_functions/thnn/activation.py
Comment (apaszke) [inline torch/nn/_functions/thnn/activation.py]: I think `grad_input = grad_output.masked_fill(input > ctx.threshold, 0)` would be more efficient.

**Review by apaszke — COMMENTED** [review]
That looks great! Thanks for the PR

**Comment by caogang** [comment]
@apaszke Thank you for your suggestions

All suggestions besides the last one `grad_input = grad_output.masked_fill(input > ctx.threshold, 0)` have been done in the new commit.

I have tried using `mask_fill` before, but I find it doesn't work when performing twice backward.

Here is my test code:
```python
def calc_gradient_penalty(netD, real_data, fake_data):
    alpha = torch.rand(BATCH_SIZE, 1)
    alpha = alpha.expand(real_data.size())
    alpha = alpha.cuda() if use_cuda else alpha

    interpolates = alpha * real_data + ((1 - alpha) * fake_data)
    if use_cuda:
        interpolates = interpolates.cuda()
    interpolates = autograd.Variable(interpolates, requires_grad=True)

    disc_interpolates = netD(interpolates)

    gradients = autograd.grad(outputs=disc_interpolates, inputs=interpolates,
                              grad_outputs=torch.ones(disc_interpolates.size()).cuda() if use_cuda else torch.ones(disc_interpolates.size()),
                              create_graph=True, only_inputs=True, retain_graph=True)[0]

    gradient_penalty = ((gradients.norm(2, dim=1) - 1) ** 2).mean() * LAMBDA
    return gradient_penalty

use_cuda = False
BATCH_SIZE=256
LAMBDA = 0.1
DIM = 256
noise = torch.neg(torch.randn(BATCH_SIZE, 2))
if use_cuda:
    noise = noise.cuda()
noisev = autograd.Variable(noise)

noise1 = torch.randn(BATCH_SIZE, 2)
if use_cuda:
    noise1 = noise1.cuda()
noise1v = autograd.Variable(noise1)

netD = nn.Sequential(
            nn.Linear(2, DIM),
            nn.ReLU(True),
            nn.Linear(DIM, DIM),
            nn.ReLU(True),
            nn.Linear(DIM, DIM),
            nn.ReLU(True),
            nn.Linear(DIM, 1),
        )
netD.zero_grad()
print netD
gp = calc_gradient_penalty(netD, noisev.data, noise1v.data)
gp.backward()
for p in netD.parameters():
    print p.grad
```

Then I get an error:
```python
NotImplementedErrorTraceback (most recent call last)
<ipython-input-2-a841c9970f31> in <module>()
     44 print netD
     45 gp = calc_gradient_penalty(netD, noisev.data, noise1v.data)
---> 46 gp.backward()
     47 for p in netD.parameters():
     48     print p.grad

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/variable.pyc in backward(self, gradient, retain_variables)
    150                 raise TypeError("gradient has to be a Tensor, Variable or None")
    151             gradient = Variable(gradient, volatile=True)
--> 152         self._execution_engine.run_backward((self,), (gradient,), retain_variables)
    153 
    154     def register_hook(self, hook):

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/function.pyc in backward(*grad_outputs)
    170         be the gradient w.r.t. the corresponding input.
    171         """
--> 172         raise NotImplementedError
    173 
    174 

NotImplementedError: 
```

I have no idea where is wrong. So I still use `grad_input = mask.type_as(grad_output) * grad_output`

**Review by apaszke — COMMENTED** [review]
I looked into that `masked_fill` issue and it turns out that it was a bug (I've included a fix in #1506). Can you please change it as I said? The tests will fail now, but should be ok once my PR is merged.

File: torch/nn/_functions/thnn/activation.py
Comment (caogang) [inline torch/nn/_functions/thnn/activation.py]: ok, I will push a commit right now.

**Comment by caogang** [comment]
Ok, changes have been done. I want to know whether the bug #1517 will be also fixed in your pull request #1506

**Comment by apaszke** [comment]
Yes, I'll push that soon as well

**Comment by apaszke** [comment]
Can you please fix the conflicts?

**Comment by caogang** [comment]
Hi @apaszke , current commit maybe still something wrong with the `mask_fill`. Because it works fine using `mask.type_as(grad_output) * grad_output`

With the same test code above. I got another error
```
TypeErrorTraceback (most recent call last)
<ipython-input-3-a841c9970f31> in <module>()
     44 print netD
     45 gp = calc_gradient_penalty(netD, noisev.data, noise1v.data)
---> 46 gp.backward()
     47 for p in netD.parameters():
     48     print p.grad

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/variable.pyc in backward(self, gradient, retain_graph, create_graph, retain_variables)
    143             Defaults to False, unless ``gradient`` is a volatile Variable.
    144         """
--> 145         torch.autograd.backward(self, gradient, retain_graph, create_graph, retain_variables)
    146 
    147     def register_hook(self, hook):

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/__init__.pyc in backward(variables, grad_variables, retain_graph, create_graph, retain_variables)
     96 
     97     Variable._execution_engine.run_backward(
---> 98         variables, grad_variables, retain_graph)
     99 
    100 

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/function.pyc in apply(self, *args)
     88 
     89     def apply(self, *args):
---> 90         return self._forward_cls.backward(self, *args)
     91 
     92 

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/nn/_functions/linear.pyc in backward(ctx, grad_output)
     21         grad_input = grad_weight = grad_bias = None
     22         if ctx.needs_input_grad[0]:
---> 23             grad_input = torch.mm(grad_output, weight)
     24         if ctx.needs_input_grad[1]:
     25             grad_weight = torch.mm(grad_output.t(), input)

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/variable.pyc in mm(self, matrix)
    526     def mm(self, matrix):
    527         output = Variable(self.data.new(self.data.size(0), matrix.data.size(1)))
--> 528         return self._static_blas(Addmm, (output, 0, 1, self, matrix), False)
    529 
    530     def bmm(self, batch):

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/variable.pyc in _static_blas(cls, args, inplace)
    519         if num_args == 4:
    520             alpha = args[1]
--> 521         return cls.apply(*(args[:1] + args[-2:] + (alpha, beta, inplace)))
    522 
    523     def _blas(self, cls, args, inplace):

/home/users/gang.cao/env/lib/python2.7/site-packages/torch/autograd/_functions/blas.pyc in forward(ctx, add_matrix, matrix1, matrix2, alpha, beta, inplace)
     22         output = _get_output(ctx, add_matrix, inplace=inplace)
     23         return torch.addmm(alpha, add_matrix, beta,
---> 24                            matrix1, matrix2, out=output)
     25 
     26     @staticmethod

TypeError: torch.addmm received an invalid combination of arguments - got (int, torch.ByteTensor, int, torch.ByteTensor, torch.FloatTensor, out=torch.ByteTensor), but expected one of:
 * (torch.ByteTensor source, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (torch.ByteTensor source, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (int beta, torch.ByteTensor source, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (torch.ByteTensor source, int alpha, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (int beta, torch.ByteTensor source, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (torch.ByteTensor source, int alpha, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (int beta, torch.ByteTensor source, int alpha, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
      didn't match because some of the arguments have invalid types: (int, torch.ByteTensor, int, torch.ByteTensor, !torch.FloatTensor!, out=torch.ByteTensor)
 * (int beta, torch.ByteTensor source, int alpha, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
      didn't match because some of the arguments have invalid types: (int, torch.ByteTensor, int, !torch.ByteTensor!, !torch.FloatTensor!, out=torch.ByteTensor)
```

**Comment by caogang** [comment]
Hi, @apaszke when will this PR be merged?

**Comment by apaszke** [comment]
I've been busy working on other things. I'll try to review it this weekend

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thank you!

**Comment by EthanZhu90** [comment]
Hi, @caogang @apaszke, I am trying the WGAN-GP implementation with pytorch installed from the latest git, but still got the same error  @caogang.  Any idea how to solve it??
Thank you in advance. 
``` python
TypeError: torch.addmm received an invalid combination of arguments - got (int, torch.ByteTensor, int, torch.ByteTensor, torch.FloatTensor, out=torch.ByteTensor), but expected one of:
 * (torch.ByteTensor source, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (torch.ByteTensor source, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (int beta, torch.ByteTensor source, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (torch.ByteTensor source, int alpha, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (int beta, torch.ByteTensor source, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (torch.ByteTensor source, int alpha, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
 * (int beta, torch.ByteTensor source, int alpha, torch.ByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
      didn't match because some of the arguments have invalid types: (int, torch.ByteTensor, int, torch.ByteTensor, !torch.FloatTensor!, out=torch.ByteTensor)
 * (int beta, torch.ByteTensor source, int alpha, torch.SparseByteTensor mat1, torch.ByteTensor mat2, *, torch.ByteTensor out)
      didn't match because some of the arguments have invalid types: (int, torch.ByteTensor, int, !torch.ByteTensor!, !torch.FloatTensor!, out=torch.ByteTensor)
```

**Comment by caogang** [comment]
@EthanZhu90 , this may be a bug existing in current branch.
So I can give you an temp fix to make this error clear **before** this bug is solved.
You can change the source code, and recompile it. This error will be clear.

```python
torch/nn/_functions/thnn/activation.py

         else:
+            mask = input > ctx.threshold
+            grad_input = mask.type_as(grad_output) * grad_output
-            grad_input = grad_output.masked_fill(input > ctx.threshold, 0)
         return grad_input, None, None, None
```

**Pull request merged**
